client/client.cpp

```diff
@@ -103,18 +103,16 @@ Client::~Client() {
 
 //============================= OPERATIONS ===================================
 
-bool Client::connect(
-    const std::string& hostname,
-    int port,
-    int timeoutMs /* = -1 */) {
+// Internal method
+bool Client::connect(Connection* conn) {
```

Comment:
This leaks memory if you try to override an active connection. What about accepting `std::unique_ptr<Connection>` in this method and do a move-assign in line 115? Alternatively this could be done with a variadic template and perfect forwarding.